packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -147,33 +225,109 @@ contract PermitERC20UniswapV3Paymaster is BasePaymaster, ERC2771Recipient {
-        emit TokensCharged(gasUseWithoutPost, gasUsedByPost, ethActualCharge);
-    }
-
-    function _depositProceedsToHub(uint256 ethActualCharge) private {
-        UniswapV3Helper.swapToEth(
-            address(weth),
-            address(token),
-            ethActualCharge,
-            uniswapPoolFee,
-            uniswap
-        );
-        relayHub.depositFor{value : address(this).balance}(address(this));
+        (address payer, uint256 priceQuote, uint256 tokenPreCharge) = abi.decode(context, (address, uint256, uint256));
+        IERC20Metadata token = _getTokenFromPaymasterData(relayData.paymasterData);
+        (uint256 tokenActualCharge, uint256 ethActualCharge) = _calculateCharge(relayData, gasUseWithoutPost + gasUsedByPost, priceQuote, tokensSwapData[token].priceDivisor);
+        require(tokenActualCharge <= tokenPreCharge, "actual charge higher");
+        token.safeTransfer(payer, tokenPreCharge - tokenActualCharge);
+
+        emit TokensCharged(gasUseWithoutPost, gasUsedByPost, tokenActualCharge, ethActualCharge);
+        _refillHubDepositIfNeeded(ethActualCharge);
+        _withdrawToOwnerIfNeeded();
+    }
+
+    function _refillHubDepositIfNeeded(uint256 ethActualCharge) private {
+        uint256 hubBalance = relayHub.balanceOf(address(this));
+        if (hubBalance >= minHubBalance + ethActualCharge) {
+            return;
+        }
+        uint256 depositAmount = targetHubBalance - hubBalance + ethActualCharge;
+        _refillHubDeposit(depositAmount);
+    }
+
+    function _refillHubDeposit(uint256 depositAmount) private {
+        uint256 balance = address(this).balance;
+        uint256 amountSwapped = 0;
+        if (balance < depositAmount) {
+            for (uint256 i = 0; i < tokens.length && balance + amountSwapped < depositAmount; i++) {
+                IERC20Metadata tokenIn = tokens[i];
+                uint256 tokenBalance = tokenIn.balanceOf(address(this));
+                if (tokenBalance > 0) {
+                    TokenSwapData memory tokenSwapData = tokensSwapData[tokenIn];
+                    uint256 quote = uint256(tokenSwapData.priceFeed.latestAnswer());
+                    uint256 amountOutMin = _tokenToWei(tokenBalance, tokenSwapData.priceDivisor, quote) * 99 / 100;
+                    uint256 amountOut = UniswapV3Helper.swapToToken(
+                        address(tokenIn),
+                        address(weth),
+                        tokenBalance,
+                        amountOutMin,
+                        tokenSwapData.uniswapPoolFee,
+                        uniswap
+                    );
+                    amountSwapped += amountOut;
+                }
+            }
+            UniswapV3Helper.unwrapWeth(uniswap, amountSwapped);
+        }
+        if (balance + amountSwapped < depositAmount) {
+            emit FundingNeeded();
+            depositAmount = balance + amountSwapped;
+        }
+        relayHub.depositFor{value : depositAmount}(address(this));
```

Comment:
TokenPaymaster should never hold ether. Can we send `balanceOf(this)`?

Reply:
Fixed